File: golang/src/chapter-9/money.go
package money

type Money interface {
	Times(m int) Money

	Currency() string
}

type moneyStructure struct {
	amount   int
	currency string
}

func (m moneyStructure) Currency() string {
	return m.currency
}

type Dollar struct {
	moneyStructure
}

func (d Dollar) Times(m int) Money {
	return NewDollar(d.amount * m)
}
func NewDollar(amount int) Money {
	return Dollar{
		moneyStructure: moneyStructure{
			amount:   amount,
			currency: "USD",
		},
	}
}

type Franc struct {
	moneyStructure
}

func (d Franc) Times(m int) Money {
	return NewFranc(d.amount * m)
}
func NewFranc(amount int) Money {
	return Franc{
		moneyStructure: moneyStructure{
			amount:   amount,
			currency: "CHF",
		},
	}
}

File: haskell/src/Money.hs
module Money where

class Expression a where
  reduce :: a -> String -> Money

data Sum = Sum { augend :: Money, addend :: Money }

instance Expression Sum where
  reduce (Sum aug add) currency = money (a1 Prelude.+ a2) currency
    where
      Money { amount = a1 } = aug
      Money { amount = a2 } = add

class Times a where
  (*) :: a -> Int -> a
class Add a where
  (+) :: a -> a -> Sum

data Money =
  Money { amount :: Int, currency :: String }
  deriving (Eq, Show)

instance Times Money where
  (Money a c) * m = Money (a Prelude.* m) c
instance Add Money where
  a + b = Sum a b
instance Expression Money where
  reduce m _ = m

money :: Int -> String -> Money
money = Money

dollar :: Int -> Money
dollar m = Money m "USD"

franc :: Int -> Money
franc m = Money m "CHF"
